nucliadb/src/nucliadb/search/search/rank_fusion.py

```diff
@@ -0,0 +1,148 @@
+    - 2 keyword results and 1 semantic (and repeat)
+
+    """
+
+    def fuse(
+        self, keyword: Iterable[TextBlockMatch], semantic: Iterable[TextBlockMatch]
+    ) -> list[TextBlockMatch]:
+        merged: list[TextBlockMatch] = []
+
+        # sort results by it's score before merging them
+        keyword = [k for k in sorted(keyword, key=lambda r: r.score, reverse=True)]
+        semantic = [s for s in sorted(semantic, key=lambda r: r.score, reverse=True)]
+
+        for k in keyword:
+            merged.append(k)
+
+        nextpos = 1
+        for s in semantic:
+            merged.insert(nextpos, s)
+            nextpos += 3
+
+        return merged
+
+
+class ReciprocalRankFusion(RankFusionAlgorithm):
+    """Rank-based rank fusion algorithm. Discounts the weight of documents
+    occurring deep in retrieved lists using a reciprocal distribution
+
+    RRF = Σ(r ∈ R) 1 / (k + r(d))
+
+    where:
+    - d is a document
+    - R is the set of retrievers
+    - k (constant)
+    - r(d) rank of document d in reranker r
+
+    """
+
+    # TODO: implement rank window
+    def __init__(self, k: float = 60.0):
+        # Constant used in RRF, studies agree on 60 as a good default value
+        # giving good results across many datasets. k allow bigger score
+        # difference among the best results and a smaller score difference among
+        # bad results
+        self.k = k
+
+    def fuse(
+        self, keyword: Iterable[TextBlockMatch], semantic: Iterable[TextBlockMatch]
+    ) -> list[TextBlockMatch]:
+        scores: dict[ParagraphId, float] = {}
+        match_positions: dict[ParagraphId, list[tuple[int, int]]] = {}
+
+        # sort results by it's score before merging them
+        keyword = [k for k in sorted(keyword, key=lambda r: r.score, reverse=True)]
+        semantic = [s for s in sorted(semantic, key=lambda r: r.score, reverse=True)]
+
+        rankings = [keyword, semantic]
+        for r, ranking in enumerate(rankings):
+            for i, result in enumerate(ranking):
+                id = result.paragraph_id
+                scores.setdefault(id, 0)
+                scores[id] += 1 / (self.k + i)
+
+                position = (r, i)
+                match_positions.setdefault(result.paragraph_id, []).append(position)
+
+        merged = []
+        for paragraph_id, positions in match_positions.items():
+            for r, i in positions:
+                score = scores[paragraph_id]
+                result = rankings[r][i]
+                result.score = score
+                result.score_type = SCORE_TYPE.RANK_FUSION
+                # NOTE we are appending multi-matches. Should we merge them?
+                merged.append(result)
+
+        merged.sort(key=lambda x: x.score, reverse=True)
+        return merged
+
+
+def get_default_rank_fusion() -> RankFusionAlgorithm:
+    return LegacyRankFusion()
+
+
+def get_rank_fusion(
+    rf: Union[search_models.RankFusionName, search_models.RankFusion],
```

Comment:
should be `Optional[Union[...]]`

Reply:
Oh, it was a leftover. I won't compare with `None` and keep the `Union` as it is now